### PR TITLE
fix(backend):Change enqueue base delay in non-error mode to 1 second for persistent agent

### LINKED ISSUE #4886: Persistence agent delays syncing workflow in the db and UI lags
### What steps did you take:
Submit a run

### What happened:
Sometimes the workflow is running (or even completes) but the UI took very long time to load it (2-5 mins)

### What did you expect to happen:
The UI to be always in sync with the run's state

### Anything else you would like to add:

The UI draws exactly what it finds in the database, but it seems like the database is empty.
It is the persistence agent who is responsible to sync the workflow (aka the run's state) in the database.

However:

* The persistence agent uses a rate limiting workqueue with an exponential delay rate limiter ([source](https://github.com/kubeflow/pipelines/blob/04d5e609f224d7be2e8bfc20a3fa7857906cae1c/backend/src/agent/persistence/worker/persistence_worker.go#L68-L69))
* The queue has a base delay of 10 seconds, maxing out to 360 seconds ([source](https://github.com/kubeflow/pipelines/blob/04d5e609f224d7be2e8bfc20a3fa7857906cae1c/backend/src/agent/persistence/worker/persistence_worker.go#L30-L35)).
  This is the mathematical equation that decides the delay:
  ```
  delay = 10 * 2^num_failures
  ```
  [Note that [this comment](https://github.com/kubeflow/pipelines/blob/04d5e609f224d7be2e8bfc20a3fa7857906cae1c/backend/src/agent/persistence/worker/persistence_worker.go#L178) is wrong]
* The number of failures is incremented every time the agent attempts to enqueue the object: [event handler](https://github.com/kubeflow/pipelines/blob/04d5e609f224d7be2e8bfc20a3fa7857906cae1c/backend/src/agent/persistence/worker/persistence_worker.go#L116), [failure during item processing](https://github.com/kubeflow/pipelines/blob/04d5e609f224d7be2e8bfc20a3fa7857906cae1c/backend/src/agent/persistence/worker/persistence_worker.go#L196)
  This is because limiter's [`When()`](https://github.com/kubernetes/client-go/blob/73fd2ddc918095ed6a4da2073b2f79e7721cf44b/util/workqueue/default_rate_limiters.go#L89-L108) method increments the failures, and it is called even when the object doesn't get actually added in the queue (e.g., if it already is in the queue):
  1. The agent calls [`AddRateLimited()`](https://github.com/kubernetes/client-go/blob/73fd2ddc918095ed6a4da2073b2f79e7721cf44b/util/workqueue/rate_limiting_queue.go#L58-L61), which calls `AddAfter(item, limiter.When(item))`
  2. [`AddAfter()`](https://github.com/kubernetes/client-go/blob/73fd2ddc918095ed6a4da2073b2f79e7721cf44b/util/workqueue/delaying_queue.go#L149-L170): adds `item` in the queue after duration `duration`
  3. [`Add()`](https://github.com/kubernetes/client-go/blob/73fd2ddc918095ed6a4da2073b2f79e7721cf44b/util/workqueue/queue.go#L107-L127): adds `item` in the queue if it's not already in it

#### So, what's happening?

Now that we know exactly when and how the delays are decided, the question is why do we get such huge delays.

Adding a few logs here and there we can see that the persistence agent has an event handler ([source](https://github.com/kubeflow/pipelines/blob/04d5e609f224d7be2e8bfc20a3fa7857906cae1c/backend/src/agent/persistence/worker/persistence_worker.go#L77-L84)). which calls [`enqueue()`](https://github.com/kubeflow/pipelines/blob/04d5e609f224d7be2e8bfc20a3fa7857906cae1c/backend/src/agent/persistence/worker/persistence_worker.go#L105-L120) when an object is added (`AddFunc`) as well as when an object is updated `UpdateFunc`).

From the [documentation](https://pkg.go.dev/k8s.io/client-go/tools/cache@v0.0.0-20191225075139-73fd2ddc9180#ResourceEventHandler) we find this:
>* OnAdd is called when an object is added.
>* OnUpdate is called when an object is modified. Note that oldObj is the
    last known state of the object-- it is possible that several changes
    were combined together, so you can't use this to see every single
    change. OnUpdate is also called when a re-list happens, and it will
    get called even if nothing changed. This is useful for periodically
    evaluating or syncing something.

This `enqueue()` method adds the workflow in the queue using `AddRateLimited()`, triggering the exponential increase.
However, it doesn't make sense to use `AddRateLimited()` since we should need the exponential delay **only when something is failing**.

Here are the logs we added (`NumRequeues` is the number of failures which are part of the computation of the exponential delay):
```
time="2020-10-27T19:41:14Z" level=info msg="========== EVENT_HANDLER.UpdateFunc =========="
time="2020-10-27T19:41:15Z" level=info msg="ENQUEUE:: Key: kubeflow-user/dog-breed-r0p46-9gkwx, NumRequeues: 0"
time="2020-10-27T19:41:15Z" level=info msg="========== EVENT_HANDLER.UpdateFunc =========="
time="2020-10-27T19:41:15Z" level=info msg="ENQUEUE:: Key: kubeflow-user/dog-breed-r0p46-9gkwx, NumRequeues: 1"
time="2020-10-27T19:41:16Z" level=info msg="========== EVENT_HANDLER.UpdateFunc =========="
time="2020-10-27T19:41:16Z" level=info msg="ENQUEUE:: Key: kubeflow-user/dog-breed-r0p46-9gkwx, NumRequeues: 2"
time="2020-10-27T19:41:17Z" level=info msg="========== EVENT_HANDLER.UpdateFunc =========="
time="2020-10-27T19:41:17Z" level=info msg="ENQUEUE:: Key: kubeflow-user/dog-breed-r0p46-9gkwx, NumRequeues: 3"
time="2020-10-27T19:41:19Z" level=info msg="========== EVENT_HANDLER.UpdateFunc =========="
time="2020-10-27T19:41:19Z" level=info msg="ENQUEUE:: Key: kubeflow-user/dog-breed-r0p46-9gkwx, NumRequeues: 4"
time="2020-10-27T19:41:21Z" level=info msg="========== EVENT_HANDLER.UpdateFunc =========="
time="2020-10-27T19:41:21Z" level=info msg="ENQUEUE:: Key: kubeflow-user/dog-breed-r0p46-9gkwx, NumRequeues: 5"
time="2020-10-27T19:41:22Z" level=info msg="========== EVENT_HANDLER.UpdateFunc =========="
time="2020-10-27T19:41:22Z" level=info msg="ENQUEUE:: Key: kubeflow-user/dog-breed-r0p46-9gkwx, NumRequeues: 6"
time="2020-10-27T19:41:25Z" level=info msg="========== EVENT_HANDLER.UpdateFunc =========="
time="2020-10-27T19:41:25Z" level=info msg="ENQUEUE:: Key: kubeflow-user/dog-breed-r0p46-9gkwx, NumRequeues: 7"
time="2020-10-27T19:41:25Z" level=info msg="PROCESS_WORK_ITEM:: Obj: kubeflow-user/dog-breed-r0p46-9gkwx, NumRequeues(obj): 8, NumRequeues(key): 8"
```

So this is what we are probably hitting:
1. The workflow `Informer` provides the agent with a workflow to process
2. The agent starts processing the workflow
3. Multiple updates on the workflow trigger the object to be requeued multiple times (incrementing the number of requeues, aka failures)
4. The processing of the workflow fails because of some transient error (Istio transport closing)
5. The agent adds the workflow back in the queue, with the exponential delay of `X` failures (`X = num_updates + 1`)

/kind bug

### PR DESCRIPTION
**Pull request by capri-xiyue**
**Description of your changes:**
Fixed #4886 
Change enqueue base delay in non-error mode to 1 second
**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

**Comment by capri-xiyue** [comment]
assign @Bobgy

**Comment by capri-xiyue** [comment]
cc @elikatsis

**Comment by capri-xiyue** [comment]
I checked the logs, the change of base delay works.
```
+ ml-pipeline-persistenceagent-6bd54697c6-kbd9v › ml-pipeline-persistenceagent
ml-pipeline-persistenceagent-6bd54697c6-kbd9v ml-pipeline-persistenceagent time="2021-01-06T05:21:39Z" level=info msg="Syncing object:: Key: kubeflow/conditional-execution-pipeline-with-exit-handler-ngmdj"
ml-pipeline-persistenceagent-6bd54697c6-kbd9v ml-pipeline-persistenceagent time="2021-01-06T05:21:40Z" level=info msg="Syncing object:: Key: kubeflow/conditional-execution-pipeline-with-exit-handler-ngmdj"
ml-pipeline-persistenceagent-6bd54697c6-kbd9v ml-pipeline-persistenceagent time="2021-01-06T05:21:41Z" level=info msg="Syncing object:: Key: kubeflow/conditional-execution-pipeline-with-exit-handler-ngmdj"
ml-pipeline-persistenceagent-6bd54697c6-kbd9v ml-pipeline-persistenceagent time="2021-01-06T05:21:43Z" level=info msg="Syncing object:: Key: kubeflow/conditional-execution-pipeline-with-exit-handler-ngmdj"
ml-pipeline-persistenceagent-6bd54697c6-kbd9v ml-pipeline-persistenceagent time="2021-01-06T05:21:45Z" level=info msg="Syncing object:: Key: kubeflow/conditional-execution-pipeline-with-exit-handler-ngmdj"
ml-pipeline-persistenceagent-6bd54697c6-kbd9v ml-pipeline-persistenceagent time="2021-01-06T05:21:46Z" level=info msg="Syncing object:: Key: kubeflow/conditional-execution-pipeline-with-exit-handler-ngmdj"
ml-pipeline-persistenceagent-6bd54697c6-kbd9v ml-pipeline-persistenceagent time="2021-01-06T05:21:47Z" level=info msg="Syncing object:: Key: kubeflow/conditional-execution-pipeline-with-exit-handler-ngmdj"
ml-pipeline-persistenceagent-6bd54697c6-kbd9v ml-pipeline-persistenceagent time="2021-01-06T05:21:48Z" level=info msg="Syncing object:: Key: kubeflow/conditional-execution-pipeline-with-exit-handler-ngmdj"
ml-pipeline-persistenceagent-6bd54697c6-kbd9v ml-pipeline-persistenceagent time="2021-01-06T05:21:49Z" level=info msg="Syncing object:: Key: kubeflow/conditional-execution-pipeline-with-exit-handler-ngmdj"
ml-pipeline-persistenceagent-6bd54697c6-kbd9v ml-pipeline-persistenceagent time="2021-01-06T05:21:50Z" level=info msg="Syncing object:: Key: kubeflow/conditional-execution-pipeline-with-exit-handler-ngmdj"

```

File: backend/src/apiserver/resource/resource_manager.go [hunk 1]
```diff
@@ -818,7 +818,15 @@ func (r *ResourceManager) ReportWorkflowResource(workflow *util.Workflow) error
...
 		// In this case the DB entry should already been created when argo workflow CR is created.
 
-		err := r.runStore.UpdateRun(runId, workflow.Condition(), workflow.FinishedAt(), workflow.ToStringForStore())
+		// If the run was Running and got terminated (activeDeadlineSeconds set to 0),
+		// ignore its condition and mark it as such
+		condition := workflow.Condition()
+		if workflow.Spec.ActiveDeadlineSeconds != nil &&
+			*workflow.Spec.ActiveDeadlineSeconds == 0 &&
+			!workflow.IsInFinalState() {
+			condition = "Terminating"
```
Comment (Bobgy) [inline backend/src/apiserver/resource/resource_manager.go:827]: nit: Is it possible using a constant that is shared also for workflow terminate API? That makes their connection much more obvious.

**Review by Bobgy — COMMENTED** [review]
Thank you @capri-xiyue and @elikatsis!
/lgtm

Left a minor comment. Also, how hard is it to unit test some of the changes here if not all?

File: backend/src/apiserver/resource/resource_manager.go (hunk 1 above)
Comment (capri-xiyue) [reply backend/src/apiserver/resource/resource_manager.go]: done

**Comment by capri-xiyue** [comment]
> Thank you @capri-xiyue and @elikatsis!
> /lgtm
> 
> Left a minor comment. Also, how hard is it to unit test some of the changes here if not all?

Usually for the enqueue logic, usually we don't write UT for that. If possible, we can write e2e to that but that can also be flaky to test enqueue behavior.

File: backend/src/apiserver/model/run.go [hunk 2]
```diff
@@ -18,6 +18,10 @@ import (
 	"strings"
 )
 
+const (
+	RunTerminatingConditions string = "Terminating"
```
Comment (Bobgy) [inline backend/src/apiserver/model/run.go:22]: The other place that should use this const is in https://github.com/kubeflow/pipelines/blob/7540ba5c3bf003758a5c2f46ccc586a0367c4dee/backend/src/apiserver/storage/run_store.go#L608

Comment (capri-xiyue) [reply backend/src/apiserver/model/run.go:22]: done

**Comment by Bobgy** [comment]
Thanks!
/approve
/lgtm

/hold
let's give @elikatsis a chance to review the code

**Comment by capri-xiyue** [comment]
@Bobgy @elikatsis 
I took a look at the code again.I think processing set source can't prevent same item from being simultaneously handled in two different workers because of the in the work queue logic since it won't throw error when you try to insert a existed item into the processing set.
Regarding the possible side effect, one thing I could image is the race condition. But since we have resync logic, I think it should be fine.
I'm not sure whether 1 second delay here is enough to make sure same item from being simultaneously handled in two different workers.

**Comment by Bobgy** [comment]
Xiyue, I don't fully understand.
I can see the comment in https://github.com/kubeflow/pipelines/blob/768317aee33d1c00a46637549fc2753aa357e0ab/backend/src/agent/persistence/worker/persistence_worker.go#L129-L130 saying two items won't be processed at the same time.

And when we get an item from the workqueue, the `Get` method first locks a mutex: https://github.com/kubernetes/client-go/blob/0d3f257a17bb46ce0d7f022e257e3cbcd01ff5cb/util/workqueue/queue.go#L148. So one item will only be got once. Did I miss anything?

**Comment by capri-xiyue** [comment]
> https://github.com/kubernetes/client-go/blob/0d3f257a17bb46ce0d7f022e257e3cbcd01ff5cb/util/workqueue/queue.go#L148

It looks like both https://github.com/kubernetes/client-go/blob/0d3f257a17bb46ce0d7f022e257e3cbcd01ff5cb/util/workqueue/queue.go#L127 and https://github.com/kubernetes/client-go/blob/0d3f257a17bb46ce0d7f022e257e3cbcd01ff5cb/util/workqueue/queue.go#L148. makes sure that same item won't  being simultaneously handled in two different workers

**Comment by elikatsis** [comment]
@capri-xiyue 
thank you for putting so much effort in this.

I've started testing this yesterday and will report early next week whether this is good enough or not. It takes me some time since it needs to occur along with transient errors.

For now, I sense it will be enough :)

**Comment by elikatsis** [comment]
It's going good up until now. Let's merge this and I'll re-open #4886 if I hit it again :) 

So LGTM and I'll let you cancel the `hold`, just in case.

**Pull request merged**